src/main/kotlin/homework6/ParallelMergeSort.kt

```diff
@@ -1,85 +1,110 @@
 package homework6
-import kotlinx.serialization.Serializable
-import kotlinx.serialization.decodeFromString
-import kotlinx.serialization.encodeToString
-import kotlinx.serialization.json.Json
-import java.io.File
 
-@Serializable
-data class TestPerThread(val arraySize: Int, val timeArray: List<Double>)
-
-@Serializable
-data class GlobalTest(val threadsNums: List<Int>, val tests: List<TestPerThread>)
+class ParallelMergeSort(val destinationArray: IntArray, val sourceArray: IntArray, initialNumberOfThreads: Int) {
+    init {
+        sort(SubArray(0, sourceArray.lastIndex), destinationArray, 0, initialNumberOfThreads)
+    }
 
-@Serializable
-data class TestData(val threadsNums: List<Int>, val arraySizes: List<Int>)
+    private data class SubArray(val left: Int, val right: Int) {
+        val size: Int
+            get() = right - left + 1
+        val middle: Int
+            get() = (right + left) / 2
+    }
 
-// data[l, m], data[m+1, r] are sorted, merge them into sorted data[l, r]
-fun merge(data: IntArray, l: Int, m: Int, r: Int) {
-    val left = IntArray(m - l + 1) { i -> data[i + l] }
-    val right = IntArray(r - m) { i -> data[i + m + 1] }
-    var i = 0
-    var j = 0
-    var k = l
-    while (i < left.size && j < right.size) {
-        if (left[i] < right[j]) {
-            data[k] = left[i++]
-        } else {
-            data[k] = right[j++]
+   private fun lowerBound(value: Int, array: IntArray, sa: SubArray): Int {
+        if (sa.left > sa.right) return sa.left
+        var l = sa.left - 1
+        var r = sa.right + 1
+        var mid: Int
+        // array[l] < value <= array[r]
+        while (l + 1 < r) {
+            mid = (l + r) / 2
+            if (array[mid] < value) {
+                l = mid
+            } else {
+                r = mid
+            }
         }
-        k++
+        return r
     }
 
-    while (i < left.size)
-        data[k++] = left[i++]
+   // for detekt to calm down on parameter list
+   private data class ArraysForMerge(val from: IntArray, val to: IntArray)
 
-    while (j < right.size)
-        data[k++] = right[j++]
-}
-
-fun sort(data: IntArray, l: Int, r: Int, numberOfThreads: Int) {
-    if (l < r) {
-        val mid = (l + r) / 2
+   // sort sourceArray[l, r] and save into destArray[s, s + r - l]
+   private fun sort(sa: SubArray, destArray: IntArray, s: Int, numberOfThreads: Int = 1) {
+        if (sourceArray.isEmpty()) return
+        if (sa.size == 1) {
+            destArray[s] = sourceArray[sa.left]
+            return
+        }
+        val tempArray = IntArray(sa.size)
+        // number of elements in the right part of subArray sa
+        val n1 = sa.middle - sa.left + 1
```

Comment:
извини, но я все равно буду придираться к именам переменных
из [naming conventions](https://kotlinlang.org/docs/coding-conventions.html#choose-good-names):

> The names should make it clear what the purpose of the entity is, so it's best to avoid using meaningless words


читатель твоего кода в идеале должен понимать, в чем смысл переменной, не обращаясь к ее инициализации (и особенно забавно, что в одном месте ты написал коммент, что это за переменная такая, когда это все должно быть понятно из имени). поэтому sa, n1, th, s, l, r и все твои остальные имена надо сделать осмысленными! используй слова целиком, используй несколько слов вместе, код должен быть понятным и в перспективе поддерживаемым долгие годы разными программистами